src/marks/density.js

```diff
@@ -11,44 +12,113 @@ const defaults = {
+            .attr("d", ([i]) => path(contours[i])))
+        .node();
   }
 }
 
 export function density(data, {x, y, ...options} = {}) {
   ([x, y] = maybeTuple(x, y));
   return new Density(data, {...options, x, y});
 }
+
+function densityInitializer(options, bandwidth, thresholds, f, s) {
+  return initializer(options, function(data, facets, channels, scales, dimensions) {
+    const X = channels.x?.scale ? valueof(channels.x.value, scales[channels.x.scale]) : channels.x?.value;
+    const Y = channels.y?.scale ? valueof(channels.y.value, scales[channels.y.scale]) : channels.y?.value;
+    const W = channels.weight?.value;
+    const Z = channels.z?.value;
+    const {z} = this;
+    const [cx, cy] = applyFrameAnchor(options, dimensions);
+    const {width, height} = dimensions;
+    const newFacets = [];
+    const contours = [];
+    const newChannels = Object.entries(channels).filter(([key]) => key !== "x" && key !== "y" && key !== "weight").map(([key, d]) => [key, {...d, value: []}]);
+    if (f) newChannels.push(["fill", {value: [], scale: "color"}]);
+    if (s) newChannels.push(["stroke", {value: [], scale: "color"}]);
+    let max = 0, maxn = 0;
+    const density = contourDensity()
+      .x(X ? i => X[i] : constant(cx))
+      .y(Y ? i => Y[i] : constant(cy))
+      .weight(W ? i => W[i] : 1)
+      .size([width, height])
+      .bandwidth(bandwidth)
+      .thresholds(thresholds);
+
+    // First pass: seek the maximum density across all facets and series; memoize for performance.
+    const memo = [];
+    thresholds = [];
+    for (const [facetIndex, facet] of facets.entries()) {
+      newFacets.push([]);
+      for (const index of Z ? groupZ(facet, Z, z) : [facet]) {
+        const c = density(index);
+        const d = c[c.length - 1];
+        if (d) {
+          if (d.value > max) {
+            max = d.value;
+            maxn = c.length;
+            thresholds = c.map(d => d.value);
+          }
+          memo.push({facetIndex, index, c, top: d.value});
+        }
+      }
+    }
+
+    // Second pass: generate contours with the thresholds derived above
+    density.thresholds(thresholds);
+    for (const {facetIndex, index, c: memoc, top} of memo) {
+      const c = top < max ? density(index) : memoc;
+      for (const contour of c) {
+        newFacets[facetIndex].push(contours.length);
+        contours.push(contour);
+        for (const [key, {value}] of newChannels) {
+          value.push(
+            (f && key === "fill") || (s && key === "stroke") ? contour.value
+            : channels[key].value[index[0]]
+          );
+        }
+      }
+    }
+
+    channels = {contours: {value: contours}, ...Object.fromEntries(newChannels)};
+    // normalize colors to a thresholds scale
+    const m = max * (maxn + 1) / maxn;
+    if (f) channels.fill.value = channels.fill.value.map(v => v / m);
+    if (s) channels.stroke.value = channels.stroke.value.map(v => v / m);
```

Comment:
I’m thinking maybe we don’t do this, and just leave the units as “points per square pixel” or “value per square pixel” or perhaps “points per 1,000 square pixels” as a more human-readable value. Even though the value is arbitrary and resolution-dependent, it’s still probably more useful than 0 = 0, 1 = maximum observed density because it allows you to have a consistent scale across charts.